components/resizable-iframe/index.js

```diff
@@ -0,0 +1,173 @@
+ */
+import { omit } from 'lodash';
+
+export default class ResizableIframe extends wp.element.Component {
+
+	constructor() {
+		super( ...arguments );
+		this.state = {
+			width: 0,
+			height: 0,
+		};
+		this.getFrameBody = this.getFrameBody.bind( this );
+		this.maybeConnect = this.maybeConnect.bind( this );
+		this.isFrameAccessible = this.isFrameAccessible.bind( this );
+		this.checkMessageForResize = this.checkMessageForResize.bind( this );
+	}
+
+	static get defaultProps() {
+		return {
+			onLoad: () => {},
+			onResize: () => {},
+			title: '',
+		};
+	}
+
+	componentDidMount() {
+		window.addEventListener( 'message', this.checkMessageForResize, false );
+		this.maybeConnect();
+	}
+
+	componentDidUpdate() {
+		this.maybeConnect();
+	}
+
+	componentWillUnmount() {
+		window.removeEventListener( 'message', this.checkMessageForResize );
+	}
+
+	getFrameBody() {
+		return this.iframe.contentDocument.body;
+	}
+
+	maybeConnect() {
+		if ( ! this.isFrameAccessible() ) {
+			return;
+		}
+
+		const body = this.getFrameBody();
+		if ( null !== body.getAttribute( 'data-resizable-iframe-connected' ) ) {
+			return;
+		}
+
+		const script = document.createElement( 'script' );
+		script.innerHTML = `
+			( function() {
+				var observer;
+
+				if ( ! window.MutationObserver || ! document.body || ! window.top ) {
+					return;
+				}
+
+				function sendResize() {
+					window.top.postMessage( {
+						action: 'resize',
+						width: document.body.offsetWidth,
+						height: document.body.offsetHeight
+					}, '*' );
+				}
+
+				observer = new MutationObserver( sendResize );
+				observer.observe( document.body, {
+					attributes: true,
+					attributeOldValue: false,
+					characterData: true,
+					characterDataOldValue: false,
+					childList: true,
+					subtree: true
+				} );
+
+				window.addEventListener( 'load', sendResize, true );
+
+				// Hack: Remove viewport unit styles, as these are relative
+				// the iframe root and interfere with our mechanism for
+				// determining the unconstrained page bounds.
+				function removeViewportStyles( ruleOrNode ) {
+					[ 'width', 'height', 'minHeight', 'maxHeight' ].forEach( function( style ) {
+						if ( /^\\d+(vmin|vmax|vh|vw)$/.test( ruleOrNode.style[ style ] ) ) {
+							ruleOrNode.style[ style ] = '';
+						}
+					} );
+				}
+
+				Array.prototype.forEach.call( document.querySelectorAll( '[style]' ), removeViewportStyles );
+				Array.prototype.forEach.call( document.styleSheets, function( stylesheet ) {
+					Array.prototype.forEach.call( stylesheet.cssRules || stylesheet.rules, removeViewportStyles );
+				} );
+
+				document.body.style.position = 'absolute';
+				document.body.setAttribute( 'data-resizable-iframe-connected', '' );
+
+				// Make sure that we don't miss very quick loading documents here that the observer
+				// doesn't see load, but haven't completely loaded when we call sendResize for the
+				// first time.
+				setTimeout( sendResize, 1000 );
```

Comment:
This works... most of the time. There's still something going on though that stops some embeds sending a resize when they load.

I'm going to tackle that specific problem in another PR. I have a good idea of how to solve it from looking at the embed code in Calypso, but I don't want to delay this and have this branch fall out of sync too much, as I'm not sure how long it will take to fully solve this.